logtracer/tracing.py

```diff
@@ -0,0 +1,122 @@
+import time
+from threading import Thread, local
+
+from google.auth.exceptions import DefaultCredentialsError
+from google.cloud.trace_v2 import TraceServiceClient
+from google.protobuf.timestamp_pb2 import Timestamp
+from google.protobuf.wrappers_pb2 import BoolValue, Int32Value
+
+from logtracer import _b3
+from logtracer import _global_vars
+from logtracer._b3 import generate_new_subspan_values
+
+SPAN_DISPLAY_NAME_BYTE_LIMIT = 128
+
+thread_memory = local()
+thread_memory.span_started = False
+trace_client = None
+
+
+class TraceException(Exception):
+    pass
+
+
+class StackDriverAuthError(Exception):
+    pass
+
+
+def configure_tracing(post_spans_to_stackdriver_api=False):
+    if post_spans_to_stackdriver_api:
+        try:
+            global trace_client
+            trace_client = TraceServiceClient()
+        except DefaultCredentialsError:
+            raise StackDriverAuthError('Cannot post spans to API, no authentication credentials found.')
+
+    _global_vars.post_spans_to_stackdriver_api = post_spans_to_stackdriver_api
+
+
+def start_traced_span(incoming_headers, request_path):
+    """
+    Start a b3 span and keep track of extra details for tracing.
+
+    Arguments:
+        incoming_headers: Incoming request headers. These could be http, or part of a GRPC message.
+        request_path (str): Path of the endpoint of the incoming request.
+    """
+    thread_memory.span = {
```

Comment:
Had you considered attaching everything to the Flask request context instead of manually doing all of this cross-thread stuff?

python-opentracing uses it and it might tidy up quite a lot of the code here if it works! https://github.com/opentracing-contrib/python-flask/blob/master/flask_opentracing/tracer.py